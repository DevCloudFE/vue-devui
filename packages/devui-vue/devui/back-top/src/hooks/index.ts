import usePosition from './use-position';
import useTarget from './use-target';
import useVisibility from './use-visibility';
import useEventListener from './use-event-listener';
import useThrottle from './use-throttle';

export { usePosition, useTarget, useVisibility, useEventListener, useThrottle };
