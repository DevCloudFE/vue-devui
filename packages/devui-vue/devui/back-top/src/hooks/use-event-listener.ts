type cbFn = () => void;
/* 简陋版待完善 */
function useEventListener(target: HTMLElement, event: string, cb: cbFn): void {
  if (target) {target.addEventListener(event, cb);}
}

export default useEventListener;
