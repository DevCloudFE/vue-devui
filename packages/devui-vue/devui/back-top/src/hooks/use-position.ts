import { BackTopProps, Position } from '../back-top-types';

export default function (props: BackTopProps): Position {
  const { bottom, right } = props;

  return {
    bottom,
    right
  };
}
