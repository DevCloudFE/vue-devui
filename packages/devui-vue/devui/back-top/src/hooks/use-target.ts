import type { BackTopProps, IBackTopRef } from '../back-top-types';

export default function (props: BackTopProps, backTopRef: IBackTopRef): Element | (Window & typeof globalThis) {
  const target = props.target; // target为元素选择器

  const getTargetEl = () => {
    const targetEl = document.querySelector(target);
    if (!targetEl) {throw new Error(`props ${target} is not existed`);}

    // 设置定位
    if (targetEl.parentElement) {
      targetEl.parentElement.style.position = 'relative';
    }
    if (backTopRef.value) {
      backTopRef.value.style.position = 'absolute';
    }
    return targetEl;
  };
  const currTarget =
    target === 'window' ? window || document.documentElement || document.body : getTargetEl();

  return currTarget;
}
