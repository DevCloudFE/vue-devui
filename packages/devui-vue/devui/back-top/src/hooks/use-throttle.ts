export default function (fn: () => void, delay: number): () => void {
  let last = 0;
  return (...args) => {
    const now = +Date.now();
    if (now - last > delay) {
      last = now;
      return fn.apply(window, args);
    }
  };
}
