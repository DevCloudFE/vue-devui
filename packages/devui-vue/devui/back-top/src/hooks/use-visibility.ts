import { onMounted, computed, ref } from 'vue';
import type { ComputedRef } from 'vue';
import { BackTopProps, IBackTopRef } from '../back-top-types';
import { useTarget, useEventListener, useThrottle } from '.';
type ICurrentTarget = HTMLElement & Window & typeof globalThis;
export default function (props: BackTopProps, backTopRef: IBackTopRef): ComputedRef<boolean> {
  const visibleHeight = props.visibleHeight;
  const currScrollTop = ref(0);
  let currTarget: ICurrentTarget;
  const ThrottleCBFn = function () {
    currScrollTop.value = currTarget === window ? window.pageYOffset : currTarget.scrollTop;
  };

  onMounted(() => {
    currTarget = useTarget(props, backTopRef) as ICurrentTarget;
    // 监听滚动事件 手动更新ScrollTop
    useEventListener(currTarget, 'scroll', useThrottle(ThrottleCBFn, 150));
  });

  const isVisible = computed(() => currScrollTop.value >= visibleHeight);

  return isVisible;
}
