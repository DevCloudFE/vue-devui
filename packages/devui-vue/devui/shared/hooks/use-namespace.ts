export type UseNamespace = {
  b: () => string;
  e: (el: string) => string;
  m: (mo: string) => string;
  em: (el: string, mo: string) => string;
};

function createBem(namespace: string, element?: string, modifier?: string): string {
  let cls = namespace;
  if (element) {
    cls += `__${element}`;
  }
  if (modifier) {
    cls += `--${modifier}`;
  }
  return cls;
}

/**
 * useNamespace
 * @param block current block name
 * @param needDot Do you need a dot prefix (defalut: false)
 * @returns UseNamespace
 */
export function useNamespace(block: string, needDot = false): UseNamespace {
  const namespace = needDot ? `.devui-${block}` : `devui-${block}`;
  const b = () => createBem(namespace);
  const e = (element: string) => (element ? createBem(namespace, element) : '');
  const m = (modifier: string) => (modifier ? createBem(namespace, '', modifier) : '');
  const em = (element: string, modifier: string) => (element && modifier ? createBem(namespace, element, modifier) : '');
  return {
    b,
    e,
    m,
    em,
  };
}
