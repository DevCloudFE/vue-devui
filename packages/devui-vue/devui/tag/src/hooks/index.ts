import useClass from './use-class';
import useColor from './use-color';

export { useClass, useColor };
