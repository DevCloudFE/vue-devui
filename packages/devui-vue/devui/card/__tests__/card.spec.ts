import { mount } from '@vue/test-utils';
import DCard from '../src/card';
import DAvatar from '../../avatar/src/avatar';
import { useNamespace } from '../../shared/hooks/use-namespace';

const ns = useNamespace('card', true);
const avatar = useNamespace('avatar');

const avatarBaseClass = avatar.b();
const titleClass = ns.e('title');
const subtitleClass = ns.e('subtitle');
const contentClass = ns.e('content');
const actionsClass = ns.e('actions');
const metaClass = ns.e('meta');
const cardBlockClass = '.card-block';
const spaceBetweenClass = `${ns.em('actions', 'align')}-spaceBetween`;
const alwaysShadowClass = ns.em('shadow', 'always');
const hoverShadowClass = ns.em('shadow', 'hover');
const neverShadowClass = ns.em('shadow', 'never');

describe('card', () => {
  it('should render correctly', async () => {
    const wrapper = mount(DCard);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should render correctly avatar', async () => {
    const wrapper = mount({
      components: {
        DCard,
        DAvatar,
      },
      template: `
        <d-card class="d-card">
          <template #avatar>
            <d-avatar name="DevUI"></d-avatar>
          </template>
        </d-card>
      `,
    });
    const avatarComponent = wrapper.findAllComponents({ name: 'dAvatar' })[0];
    expect(avatarComponent.classes()).toContain(avatarBaseClass);
  });

  it('in v-slot mode should render correctly avatar', async () => {
    const wrapper = mount({
      components: {
        DCard,
        DAvatar,
      },
      template: `
        <d-card class="d-card">
          <template v-slot:avatar>
            <d-avatar name="DevUI"></d-avatar>
          </template>
        </d-card>
      `,
    });
    const avatarComponent = wrapper.findAllComponents({ name: 'dAvatar' })[0];
    expect(avatarComponent.classes()).toContain(avatarBaseClass);
  });

  it('should render correctly title', async () => {
    const wrapper = mount({
      components: {
        DCard,
      },
      template: `
        <d-card class="d-card">
          <template #title>
            DevUI
          </template>
        </d-card>
      `,
    });
    expect(wrapper.find(titleClass).text()).toBe('DevUI');
  });

  it('in v-slot mode should render title', async () => {
    const wrapper = mount({
      components: {
        DCard,
      },
      template: `
        <d-card class="d-card">
          <template v-slot:title>
            DEVUI Course
          </template>
        </d-card>
      `,
    });
    expect(wrapper.find(titleClass).text()).toBe('DEVUI Course');
  });

  it('should render correctly subtitle', async () => {
    const wrapper = mount({
      components: {
        DCard,
      },
      template: `
        <d-card class="d-card">
          <template #subtitle>
            DevUI
          </template>
        </d-card>
      `,
    });
    expect(wrapper.find(subtitleClass).text()).toBe('DevUI');
  });

  it('in v-slot mode should render subtitle', async () => {
    const wrapper = mount({
      components: {
        DCard,
      },
      template: `
        <d-card class="d-card">
          <template v-slot:subtitle>
            DevUI
          </template>
        </d-card>
      `,
    });
    expect(wrapper.find(subtitleClass).text()).toBe('DevUI');
  });

  it('should render correctly content', async () => {
    const wrapper = mount({
      components: {
        DCard,
      },
      template: `
        <d-card class="d-card">
          <template #content>
            DevUI
          </template>
        </d-card>
      `,
    });
    expect(wrapper.find(contentClass).text()).toBe('DevUI');
  });

  it('in v-slot mode should render content', async () => {
    const wrapper = mount({
      components: {
        DCard,
      },
      template: `
        <d-card class="d-card">
          <template v-slot:content>
            DevUI
          </template>
        </d-card>
      `,
    });
    expect(wrapper.find(contentClass).text()).toBe('DevUI');
  });

  it('should render correctly actions', async () => {
    const wrapper = mount({
      components: {
        DCard,
      },
      template: `
        <d-card class="d-card">
          <template #actions>
            <div class="card-block">
              btn
            </div>
          </template>
        </d-card>
      `,
    });
    expect(wrapper.find(actionsClass).exists()).toBeTruthy();
    expect(wrapper.find(cardBlockClass).text()).toBe('btn');
  });

  it('in v-slot mode should render actions', async () => {
    const wrapper = mount({
      components: {
        DCard,
      },
      template: `
        <d-card class="d-card">
          <template #actions>
            <div class="card-block">
              btn
            </div>
          </template>
        </d-card>
      `,
    });
    expect(wrapper.find(actionsClass).exists()).toBeTruthy();
    expect(wrapper.find(cardBlockClass).text()).toBe('btn');
  });

  it('should render correctly image', async () => {
    const wrapper = mount({
      components: {
        DCard,
      },
      template: `
        <d-card class="d-card" :src="'https://devui.design/components/assets/image1.png'">
        </d-card>
      `,
    });
    expect(wrapper.find(metaClass).attributes('src').includes('https://devui.design/components/assets/image1.png')).toBeTruthy();
  });

  it('should render correctly align', async () => {
    const wrapper = mount({
      components: {
        DCard,
      },
      template: `
        <d-card class="d-card" :align="'spaceBetween'">
        </d-card>
      `,
    });
    expect(wrapper.find(spaceBetweenClass).exists()).toBeTruthy();
  });

  it('should render correctly shadow', async () => {
    const wrapper = mount({
      components: {
        DCard,
      },
      template: `
        <d-card >
        </d-card>
      `,
    });
    expect(wrapper.find(hoverShadowClass).exists()).toBeTruthy();

    await wrapper.setProps({
      shadow: 'always',
    });
    expect(wrapper.find(alwaysShadowClass).exists()).toBeTruthy();

    await wrapper.setProps({
      shadow: 'never',
    });
    expect(wrapper.find(neverShadowClass).exists()).toBeTruthy();
  });
});
