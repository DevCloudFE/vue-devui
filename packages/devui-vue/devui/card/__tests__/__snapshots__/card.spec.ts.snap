// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`card should render correctly 1`] = `
<div
  class="card-container devui-card devui-card__shadow--hover"
>
  <!---->
  <div
    class="devui-card__header"
  >
    
    <div>
      <div
        class="devui-card__title"
      >
        <!---->
      </div>
      <div
        class="devui-card__subtitle"
      >
        <!---->
      </div>
    </div>
  </div>
  
  <div
    class="devui-card__content"
  >
    <!---->
  </div>
  <div
    class="devui-card__actions"
  >
    
  </div>
</div>
`;
